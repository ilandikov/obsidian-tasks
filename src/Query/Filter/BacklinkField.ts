import { Group } from '../Group';
import type { Task } from '../../Task';
import { TextField } from './TextField';
import { HeadingField } from './HeadingField';

export class BacklinkField extends TextField {
    public fieldName(): string {
        return 'backlink';
    }

    public value(task: Task): string {
        const linkText = task.getLinkText({ isFilenameUnique: true });
        if (linkText === null) {
            return 'Unknown Location';
        }

        let filenameComponent = 'Unknown Location';

        if (task.filename !== null) {
            // Markdown characters in the file name must be escaped.
            filenameComponent = Group.escapeMarkdownCharacters(task.filename);
        }

        if (task.precedingHeader === null || task.precedingHeader.length === 0) {
            return filenameComponent;
        }

        // Markdown characters in the heading must NOT be escaped.
        const headingGrouper = new HeadingField().createGrouper().grouper;
        const headingComponent = headingGrouper(task)[0];

        if (filenameComponent === headingComponent) {
            return filenameComponent;
        } else {
            return `${filenameComponent} > ${headingComponent}`;
        }
    }

    public supportsGrouping(): boolean {
        return true;
    }
}
