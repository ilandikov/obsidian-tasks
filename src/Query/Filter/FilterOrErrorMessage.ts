import { QueryComponentOrError } from '../QueryComponentOrError';
import type { Filter, FilterFunction } from './Filter';

/**
 * A class which stores one of:
 * - The original instruction string - a line from a tasks code block
 * - An optional {@link Filter}
 * - An optional error message
 *
 * This is really currently a convenience for returning data from
 * {@link Field.createFilterOrErrorMessage()} and derived classes.
 *
 * By the time the code has finished with parsing the line, typically the
 * contained {@link Filter} will be saved, for later use in searching for Tasks
 * that match the user's filter instruction.
 */
export class FilterOrErrorMessage extends QueryComponentOrError<Filter> {
    public get filter(): Filter | undefined {
        return this.queryComponent;
    }

    get filterFunction(): FilterFunction | undefined {
        if (this.filter) {
            return this.filter.filterFunction;
        } else {
            return undefined;
        }
    }

    /**
     * Construct a FilterOrErrorMessage with the filter.
     *
     * This function allows a meaningful {@link Explanation} to be supplied.
     *
     * @param filter - a {@link Filter}
     */
    public static fromFilter(filter: Filter): FilterOrErrorMessage {
        const newFilter = new FilterOrErrorMessage(filter.instruction);
        newFilter.queryComponent = filter;
        return newFilter;
    }

    /**
     * Construct a FilterOrErrorMessage with the given error message.
     * @param instruction
     * @param errorMessage
     */
    public static fromError(instruction: string, errorMessage: string): FilterOrErrorMessage {
        const fromError = QueryComponentOrError.fromError<Filter>(instruction, errorMessage);
        return fromError as FilterOrErrorMessage;
    }
}
