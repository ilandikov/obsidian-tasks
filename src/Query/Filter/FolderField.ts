import { Group } from '../Group';
import type { Task } from '../../Task';
import { TextField } from './TextField';

export class FolderField extends TextField {
    public fieldName(): string {
        return 'folder';
    }

    public value(task: Task): string {
        const path = task.path;
        const fileNameWithExtension = task.filename + '.md';
        const folder = path.substring(0, path.lastIndexOf(fileNameWithExtension));
        if (folder === '') {
            return '/';
        }
        return Group.escapeMarkdownCharacters(folder);
    }

    public supportsGrouping(): boolean {
        return true;
    }
}
