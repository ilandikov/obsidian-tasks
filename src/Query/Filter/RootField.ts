import { Group } from '../Group';
import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { TextField } from './TextField';

export class RootField extends TextField {
    public fieldName(): string {
        return 'root';
    }

    public value(task: Task): string {
        const path = task.path.replace(/\\/g, '/');
        const separatorIndex = path.indexOf('/');
        if (separatorIndex == -1) {
            return '/';
        }
        return path.substring(0, separatorIndex + 1);
    }

    public supportsGrouping(): boolean {
        return true;
    }

    public grouper(): GrouperFunction {
        return (task: Task) => {
            const path = task.path.replace(/\\/g, '/');
            const separatorIndex = path.indexOf('/');
            if (separatorIndex == -1) {
                return ['/'];
            }
            return [Group.escapeMarkdownCharacters(path.substring(0, separatorIndex + 1))];
        };
    }
}
