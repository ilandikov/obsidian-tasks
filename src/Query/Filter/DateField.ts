import type { Moment } from 'moment';
import type { Task } from '../../Task';
import { DateParser } from '../DateParser';
import { Explanation } from '../Explain/Explanation';
import type { Comparator } from '../Sorter';
import { compareByDate } from '../../lib/DateTools';
import { Field } from './Field';
import { Filter, FilterOrErrorMessage } from './Filter';
import { FilterInstructions } from './FilterInstructions';

/**
 * DateField is an abstract base class to help implement
 * all the filter instructions that act on a single type of date
 * value, such as the done date.
 */
export abstract class DateField extends Field {
    private readonly filterInstructions: FilterInstructions;

    // TODO add previous and next
    public static readonly currentPeriodRegexp = /^current (week|month|year)$/;

    constructor() {
        super();
        this.filterInstructions = new FilterInstructions();
        this.filterInstructions.add(`has ${this.fieldName()} date`, (task: Task) => this.date(task) !== null);
        this.filterInstructions.add(`no ${this.fieldName()} date`, (task: Task) => this.date(task) === null);
        this.filterInstructions.add(`${this.fieldName()} date is invalid`, (task: Task) => {
            const date = this.date(task);
            return date !== null && !date.isValid();
        });
    }

    public canCreateFilterForLine(line: string): boolean {
        if (this.filterInstructions.canCreateFilterForLine(line)) {
            return true;
        }

        return super.canCreateFilterForLine(line);
    }

    public createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        const filterResult = this.filterInstructions.createFilterOrErrorMessage(line);
        if (filterResult.filter !== undefined) {
            return filterResult;
        }

        const result = new FilterOrErrorMessage(line);

        const match = Field.getMatch(this.filterRegExp(), line);
        let filterFunction;
        if (match !== null) {
            const filterDate = DateParser.parseDate(match[2]);
            const matchCurrentPeriod = match[2].match(DateField.currentPeriodRegexp);

            // Something is wrong if the date is wrong AND we are not in current w/m/y case
            if (!filterDate.isValid() && matchCurrentPeriod == null) {
                result.error = 'do not understand ' + this.fieldName() + ' date';
            } else {
                let relative;
                if (match[1] === 'before') {
                    filterFunction = (task: Task) => {
                        const date = this.date(task);
                        return date ? date.isBefore(filterDate) : this.filterResultIfFieldMissing();
                    };
                    relative = ' ' + match[1];
                } else if (match[1] === 'after') {
                    filterFunction = (task: Task) => {
                        const date = this.date(task);
                        return date ? date.isAfter(filterDate) : this.filterResultIfFieldMissing();
                    };
                    relative = ' ' + match[1];
                } else {
                    filterFunction = (task: Task) => {
                        const date = this.date(task);
                        return date ? date.isSame(filterDate) : this.filterResultIfFieldMissing();
                    };
                    relative = ' on';
                }

                if (match[1] === 'in') {
                    if (matchCurrentPeriod === null) {
                        result.error = 'do not understand query filter (' + this.fieldName() + ' date)';
                    } else {
                        filterFunction = (task: Task) => {
                            const date = this.date(task);
                            return date
                                ? DateField.isDateInCurrentPeriod(date, matchCurrentPeriod[1])
                                : this.filterResultIfFieldMissing();
                        };

                        result.filter = new Filter(
                            line,
                            filterFunction,
                            new Explanation(
                                DateField.getExplanationString(
                                    this.fieldName(),
                                    '',
                                    this.filterResultIfFieldMissing(),
                                    DateField.currentPeriodBoundaryDates(matchCurrentPeriod[1]),
                                ),
                            ),
                        );

                        // Exit here to return the result for 'current period'
                        return result;
                    }
                }

                const explanation = DateField.getExplanationString(
                    this.fieldName(),
                    relative,
                    this.filterResultIfFieldMissing(),
                    filterDate,
                );
                result.filter = new Filter(line, filterFunction, new Explanation(explanation));
            }
        } else {
            result.error = 'do not understand query filter (' + this.fieldName() + ' date)';
        }
        return result;
    }

    /**
     * Return the task's value for this date field, if any.
     * @param task - a Task object
     * @public
     */
    public abstract date(task: Task): Moment | null;

    /**
     * Construct a string used to explain a date-based filter
     * @param fieldName - for example, 'due'
     * @param relationshipPrefixedWithSpace - for example ' before' or ''
     * @param filterResultIfFieldMissing - whether the search matches tasks without the requested date value
     * @param filterDate - the date used in the filter
     */
    public static getExplanationString(
        fieldName: string,
        relationshipPrefixedWithSpace: string,
        filterResultIfFieldMissing: boolean,
        filterDateOrDateRange: moment.Moment | [moment.Moment, moment.Moment],
    ) {
        // Example of formatted date: '2024-01-02 (Tuesday 2nd January 2024)'
        const outputDateFormat = 'YYYY-MM-DD (dddd Do MMMM YYYY)';
        let result = `${fieldName} date is${relationshipPrefixedWithSpace}`;

        if (window.moment.isMoment(filterDateOrDateRange)) {
            const actualDate = filterDateOrDateRange.format(outputDateFormat);
            result += ` ${actualDate}`;
        } else {
            result += ' between ';
            result += filterDateOrDateRange[0].format(outputDateFormat);
            result += ' and ';
            result += filterDateOrDateRange[1].format(outputDateFormat);
            result += ' inclusive';
        }

        if (filterResultIfFieldMissing) {
            result += ` OR no ${fieldName} date`;
        }
        return result;
    }

    /**
     * Determine whether a task that does not have the particular date value
     * should be treated as a match. For example, 'starts' searches match all tasks
     * that have no start date, which behaves differently from 'due', 'done' and
     * 'scheduled' searches.
     * @protected
     */
    protected abstract filterResultIfFieldMissing(): boolean;

    public supportsSorting(): boolean {
        return true;
    }

    public comparator(): Comparator {
        return (a: Task, b: Task) => {
            return compareByDate(this.date(a), this.date(b));
        };
    }

    public static currentPeriodBoundaryDates(period: string): [moment.Moment, moment.Moment] {
        switch (period) {
            case 'week':
                // Use locale-independant ISO 8601 weeks
                return [window.moment().startOf('isoWeek'), window.moment().endOf('isoWeek')];
            case 'month':
            case 'year':
                return [window.moment().startOf(period), window.moment().endOf(period)];
        }

        // error case here?
        return [window.moment(), window.moment()];
    }
    public static isDateInCurrentPeriod(date: moment.Moment, period: string): boolean {
        const currentPeriod = DateField.currentPeriodBoundaryDates(period);

        return date.isSameOrAfter(currentPeriod[0]) && date.isSameOrBefore(currentPeriod[1]);
    }
}
