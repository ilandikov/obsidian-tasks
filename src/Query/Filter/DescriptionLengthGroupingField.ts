import type { Task } from 'Task';
import type { GrouperFunction } from 'Query/Grouper';
import { Field } from './Field';
import { FilterOrErrorMessage } from './Filter';

/** This is a class for test purposes of a Field that supports grouping but not sorting
 */
export class DescriptionLengthGroupingfield extends Field {
    protected filterRegExp(): RegExp | null {
        throw new Error('No filtering for description length field');
    }
    public fieldName(): string {
        return 'description length';
    }

    public value(task: Task): number {
        return task.description.length;
    }

    public createFilterOrErrorMessage(line: string): FilterOrErrorMessage {
        return FilterOrErrorMessage.fromError(line, 'description length field does not support filtering');
    }

    // -----------------------------------------------------------------------------------------------------------------
    // Sorting
    // -----------------------------------------------------------------------------------------------------------------

    // Doesn't support sorting by default as in Field.ts

    // -----------------------------------------------------------------------------------------------------------------
    // Grouping
    // -----------------------------------------------------------------------------------------------------------------

    public supportsGrouping(): boolean {
        return true;
    }

    public grouper(): GrouperFunction {
        return (task: Task) => {
            return [this.value(task).toString()];
        };
    }
}
