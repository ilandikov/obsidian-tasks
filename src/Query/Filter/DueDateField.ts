import type { Moment } from 'moment';
import type { Task } from '../../Task';
import type { GrouperFunction } from '../Grouper';
import { DateField } from './DateField';

/**
 * Support the 'due' search instruction.
 */
export class DueDateField extends DateField {
    public fieldName(): string {
        return 'due';
    }
    public date(task: Task): Moment | null {
        return task.dueDate;
    }
    protected filterResultIfFieldMissing() {
        return false;
    }

    public supportsGrouping(): boolean {
        return true;
    }

    public grouper(): GrouperFunction {
        return (task: Task) => {
            const date = this.date(task);
            if (date === null) {
                return ['No ' + this.fieldName() + ' date'];
            }
            return [date.format('YYYY-MM-DD dddd')];
        };
    }
}
