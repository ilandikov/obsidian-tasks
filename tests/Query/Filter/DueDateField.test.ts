/**
 * @jest-environment jsdom
 */
import moment from 'moment';
import { DueDateField } from '../../../src/Query/Filter/DueDateField';
import type { FilterOrErrorMessage } from '../../../src/Query/Filter/Filter';
import { TaskBuilder } from '../../TestingTools/TaskBuilder';
import { testFilter } from '../../TestingTools/FilterTestHelpers';
import { toHaveExplanation } from '../../CustomMatchers/CustomMatchersForFilters';
import {
    expectTaskComparesAfter,
    expectTaskComparesBefore,
    expectTaskComparesEqual,
} from '../../CustomMatchers/CustomMatchersForSorting';

window.moment = moment;

expect.extend({
    toHaveExplanation,
});

function testTaskFilterForTaskWithDueDate(filter: FilterOrErrorMessage, dueDate: string | null, expected: boolean) {
    const builder = new TaskBuilder();
    testFilter(filter, builder.dueDate(dueDate), expected);
}

describe('due date', () => {
    it('by due date (before)', () => {
        // Arrange
        const filter = new DueDateField().createFilterOrErrorMessage('due before 2022-04-20');

        // Act, Assert
        testTaskFilterForTaskWithDueDate(filter, null, false);
        testTaskFilterForTaskWithDueDate(filter, '2022-04-15', true);
        testTaskFilterForTaskWithDueDate(filter, '2022-04-20', false);
        testTaskFilterForTaskWithDueDate(filter, '2022-04-25', false);
    });

    it('due date is invalid', () => {
        // Arrange
        const filter = new DueDateField().createFilterOrErrorMessage('due date is invalid');

        // Act, Assert
        testTaskFilterForTaskWithDueDate(filter, null, false);
        testTaskFilterForTaskWithDueDate(filter, '2022-04-15', false);
        testTaskFilterForTaskWithDueDate(filter, '2022-02-30', true); // 30 February is not valid
        testTaskFilterForTaskWithDueDate(filter, '2022-00-01', true); // month 0 not valid
        testTaskFilterForTaskWithDueDate(filter, '2022-13-01', true); // month 13 not valid
    });

    it.each([
        // Week
        ['due in current week', '2022-01-09 (Sunday 9th January 2022)', false],
        ['due in current week', '2022-01-10 (Monday 10th January 2022)', true],
        ['due in current week', '2022-01-16 (Sunday 16th January 2022)', true],
        ['due in current week', '2022-01-17 (Monday 16th January 2022)', false],

        // Month
        ['due in current month', '2021-12-31 (Friday 31st December 2021)', false],
        ['due in current month', '2022-01-01 (Saturday 1st January 2022)', true],
        ['due in current month', '2022-01-31 (Monday 31st January 2022)', true],
        ['due in current month', '2022-02-01 (Tuesday 1st February 2022)', false],

        // Year
        ['due in current year', '2021-12-31 (Friday 31st December 2021)', false],
        ['due in current year', '2022-01-01 (Saturday 1st January 2022)', true],
        ['due in current year', '2022-12-31 (Saturday 31st December 2022)', true],
        ['due in current year', '2023-01-01 (Sunday 1st January 2023)', false],
    ])(
        '"%s" expect a task with "%s" date in due field to be "%s"',
        (filterString: string, testDate: string, expected: boolean) => {
            jest.useFakeTimers();
            jest.setSystemTime(new Date(2022, 0, 15)); // 2022-01-15

            const filter = new DueDateField().createFilterOrErrorMessage(filterString);

            testFilter(filter, new TaskBuilder().dueDate(null), false);
            testFilter(filter, new TaskBuilder().dueDate(testDate), expected);

            jest.useRealTimers();
        },
    );
});

describe('explain due date queries', () => {
    it('should explain explicit date', () => {
        const filterOrMessage = new DueDateField().createFilterOrErrorMessage('due before 2023-01-02');
        expect(filterOrMessage).toHaveExplanation('due date is before 2023-01-02 (Monday 2nd January 2023)');
    });

    it('implicit "on" gets added to explanation', () => {
        const filterOrMessage = new DueDateField().createFilterOrErrorMessage('due 2023-01-02');
        expect(filterOrMessage).toHaveExplanation('due date is on 2023-01-02 (Monday 2nd January 2023)');
    });

    it.each([
        [
            'due in current week',
            'due date is between 2022-01-10 (Monday 10th January 2022) and 2022-01-16 (Sunday 16th January 2022) inclusive',
        ],
        [
            'due in current month',
            'due date is between 2022-01-01 (Saturday 1st January 2022) and 2022-01-31 (Monday 31st January 2022) inclusive',
        ],
        [
            'due in current year',
            'due date is between 2022-01-01 (Saturday 1st January 2022) and 2022-12-31 (Saturday 31st December 2022) inclusive',
        ],
    ])('explains "%s" as "%s"', (filter: string, expectedExpanation: string) => {
        jest.useFakeTimers();
        jest.setSystemTime(new Date(2022, 0, 15)); // 2022-01-15

        const filterOrMessage = new DueDateField().createFilterOrErrorMessage(filter);
        expect(filterOrMessage).toHaveExplanation(expectedExpanation);

        jest.useRealTimers();
    });
});

describe('sorting by due', () => {
    const date1 = new TaskBuilder().dueDate('2021-01-12').build();
    const date2 = new TaskBuilder().dueDate('2022-12-23').build();

    it('supports Field sorting methods correctly', () => {
        const field = new DueDateField();
        expect(field.supportsSorting()).toEqual(true);
    });

    it('sort by due', () => {
        // Arrange
        const sorter = new DueDateField().createNormalSorter();

        // Assert
        expectTaskComparesBefore(sorter, date1, date2);
        expectTaskComparesAfter(sorter, date2, date1);
        expectTaskComparesEqual(sorter, date2, date2);
    });

    it('sort by due reverse', () => {
        // Arrange
        const sorter = new DueDateField().createReverseSorter();

        // Assert
        expectTaskComparesAfter(sorter, date1, date2);
        expectTaskComparesBefore(sorter, date2, date1);
        expectTaskComparesEqual(sorter, date2, date2);
    });
});
