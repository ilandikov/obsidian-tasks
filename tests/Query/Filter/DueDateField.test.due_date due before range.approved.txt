due before 2023-02-07 2023-02-11

Before the range (2023-02-06) is in
Start of the range (2023-02-07) is out
Middle of the range (2023-02-10) is out
End of the range (2023-02-11) is out
After the range (2023-02-12) is out
